### ISSUE TITLE
Distribution outside of the Mac App Store: Please remove prefix "Developer ID Application:"

### ISSUE THREAD

**Issue by mahnunchik** [electron-userland/electron-builder#4935]
* **Version**: 22.6.0

* **Electron Version**: 8.2.5
* **Electron Type (current, beta, nightly)**: current

* **Target**: dmg (Distribution outside of the Mac App Store)

I'd like to build `dmg` target to distribute app outside of App Store.

I've read the documentation:

* https://www.electron.build/code-signing
* Notarize app for macOS https://github.com/electron-userland/electron-builder/issues/3870
* https://developer.apple.com/developer-id/

My config:

```
  "build": {
    "mac": {
      "target": [
        "dmg"
      ],
    },
  }
```

I'm using the following command:

```bash
 CSC_NAME="Developer ID Application: Company Name (12345)" electron-builder --mac
```

I've faced with the following error: ` ⨯ Please remove prefix "Developer ID Application:" from the specified name — appropriate certificate will be chosen automatically`

According to the Apple documentation I must use `Developer ID certificate` https://developer.apple.com/developer-id/

If I've removed `Developer ID Application:` from `CSC_NAME` then `Apple Distribution: Company Name (12345)` was chosen.

How should I specify a certificate name to get around the error above and build `outside of App Store dmg` with the correct certificate to be able to notarize the app?

**Comment by u83ism** [comment]
I have the same problem. 
It's not good to generate p12 files unnecessarily.

**Comment by mahnunchik** [comment]
yep!

**Comment by mahnunchik** [comment]
пт, 25 сент. 2020 г., 14:04 stale[bot] <notifications@github.com>:

**Comment by alibosworth** [comment]
I'm experiencing this too

**Comment by absoftware** [comment]
I have the same problem. Please help how to force correct certificate.

**Comment by Padge91** [comment]
I may have found a workaround. Unfortunately I'm pretty unfamiliar with the Apple signing/publishing process so I'm unsure if my steps are repeatable or even reliable.

My first attempt at a solution was to set the "identity" parameter in the mac electron-builder configurations to null as suggested here https://www.electron.build/code-signing. This initially appeared to disable all signing correctly. My idea was to then perform the signing myself. After signing the multiple binaries included in Electron, my "codesign -vv --deep ../dist/mas/ExampleApp.app/" command no longer succeeded. I may have missed something, but it was warning about changed signatures so I backed off. After this I was planning to do the notarization manually, which I'm already having to do for my PKG file so it wouldn't have been a big deal. Eventually the "identity" parameter appeared to not work correctly anymore, and electron-builder tried to use the incorrect certificate in my keychain once again. At this point I gave up on this thread.

My second attempt involved me setting the "identity" parameter to the identity hash of the key I wanted electron-builder to use. This correctly uses the key I want, but then electron-builder STILL tries to use another key from keychain, at which point it fails. I'm not sure how it's trying to figure out which keys to use. Luckily, my afterSign notarize.js script still runs, so the .app still notarizes (and successfully notarizes).

Maybe this info can help someone else proceed. I'm trying to get electron-builder to generate a PKG using the correct key, which it isn't doing, so I have to do some more work. Maybe the DMG process will complete successfully, or my input above can helps someone get 95% of the way there.

Good luck!

**Comment by mahnunchik** [comment]
NOOOOO!!!

**Comment by gschafra** [comment]
We're also experiencing these issue(s)! Please help!

**Comment by ejain** [comment]
I just ran into this; the issue was that while I had a "Developer ID Application" certificate in my keychain, that didn't include the private key for the certificate... Once I added the private key to the keychain, electron-builder picked up the correct certificate, and notarization succeeded.

**Comment by ifedapoolarewaju** [comment]
Ok, so this issue cost me considerable amount of time and I mostly felt helpless until I eventually figured out the solution. This should probably work for everyone who face this issue.

When I renewed my Apple Developer membership, I found I already had two certificates under my account, so I downloaded those and added them to my Keychain. However, Electron-builder refused to pick them up during app packaging. It always picked other irrelevant certificates instead.

What I then did was to [create new certificates](https://developer.apple.com/help/account/create-certificates/create-developer-id-certificates/) and downloaded those. Once I added them to Keychain, and re-ran Electron-builder, it immediately picked them up and signed my app correctly.

What was the difference between these new certificates and the old ones? Well, one prominent thing is that these new certificates contained their `private key` (just as [mentioned](https://github.com/electron-userland/electron-builder/issues/4935#issuecomment-846311686) by @ejain ).

So it must mean the first two certificates I had downloaded were created with my old computer (some years back) and so their private keys were not present when I downloaded them from devevloper.apple.com. When creating the new certs, the Form prompts you to link a `certificate singing request file` (don't worry it's all in the instructions from the first link I shared), which I believe contains the missing private key that my older certificates didn't have.
